controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1153,6 +1180,245 @@ func saveHostProvisioningSettings(host *metal3v1alpha1.BareMetalHost) (dirty boo
 	return
 }
 
+// Get the stored firmware settings and validate the Spec settings against the schema
+func (r *BareMetalHostReconciler) getHostFirmwareSettings(prov provisioner.Provisioner, info *reconcileInfo) (hfs *metal3v1alpha1.HostFirmwareSettings, err error) {
+
+	hfs = &metal3v1alpha1.HostFirmwareSettings{}
+	if r.Get(context.TODO(), info.request.NamespacedName, hfs) != nil {
+
+		if !k8serrors.IsNotFound(err) {
+			// Error reading the object
+			return nil, errors.Wrap(err, "could not load host firmware settings")
+		}
+
+		// Could not get settings, don't return error as settings may not have been available at provisioner
+		info.log.Info("could not get hostFirmwareSettings", "namespacename", info.request.NamespacedName)
+		return nil, nil
+	}
+
+	if hfs.Status.Settings == nil {
+		// If settings cleared after pivot, restore the settings
+		currentSettings, _, err := prov.GetFirmwareSettings(false)
+		if err != nil {
+			info.log.Info("could not get firmware settings from provisioner")
+			return nil, err
+		}
+
+		if err = r.updateHostFirmwareSettingsFromCurrent(currentSettings, hfs, nil); err != nil {
+			return nil, err
+		}
+	}
+
+	schema := &metal3v1alpha1.FirmwareSchema{}
+	if hfs.Status.FirmwareSchema != nil {
+		if r.Get(context.TODO(), client.ObjectKey{
+			Namespace: hfs.Status.FirmwareSchema.Namespace,
+			Name:      hfs.Status.FirmwareSchema.Name}, schema) != nil {
+			info.log.Info("failed to get schema for hostFirmwareSettings", "namespacename", hfs.Status.FirmwareSchema.Namespace, "name", hfs.Status.FirmwareSchema.Name)
+			return hfs, nil
```

Comment:
Is it a normal situation? I would assume that if we have a `FirmwareSchema` link it must resolve? Or do you expect a race here?

Reply:
Changed this return an error.